### PR TITLE
Caplin: Optimization and Parallelization of processes and reduction of Goroutines

### PR DESCRIPTION
**Pull request by Giulio2002**
Optimizations:

1) Single goroutine for tracking expiry of gossip subscriptions
2) tweaking of parameters on operations retention in cache
3) moving the dumping of BeaconState after forkchoice, so that we do not mess up block times with I/O

### CODE REVIEW

**Comment by Giulio2002** [comment]
@domiwei feel free to review now

File: cl/sentinel/gossip.go [hunk 1]
```diff
@@ -495,6 +494,24 @@ func (g *GossipManager) Close() {
 	})
 }
 
+func (g *GossipManager) Start(ctx context.Context) {
```
Comment (domiwei) [inline cl/sentinel/gossip.go:497]: I feel the naming could be more clear like `MonitorTopicSubscriptions` or anything else

File: cl/sentinel/gossip.go [hunk 2]
```diff
@@ -208,11 +208,10 @@ func (s *Sentinel) forkWatcher() {
 				s.subManager.subscriptions.Range(func(key, value interface{}) bool {
 					sub := value.(*GossipSubscription)
 					s.subManager.unsubscribe(key.(string))
-					newSub, err := s.SubscribeGossip(sub.gossip_topic, sub.expiration.Load().(time.Time))
+					_, err := s.SubscribeGossip(sub.gossip_topic, sub.expiration.Load().(time.Time))
```
Comment (domiwei) [inline cl/sentinel/gossip.go:211]: maybe lets remove first returned value `*GossipSubscription` from `SubscribeGossip`?

File: cl/phase1/network/services/constants.go [hunk 3]
```diff
@@ -29,8 +29,8 @@ const (
 	blobJobsIntervalTick          = 5 * time.Millisecond
 	singleAttestationIntervalTick = 10 * time.Millisecond
 	attestationJobsIntervalTick   = 100 * time.Millisecond
-	blockJobExpiry                = 7 * time.Minute
-	blobJobExpiry                 = 7 * time.Minute
+	blockJobExpiry                = 24 * time.Second
```
Comment (domiwei) [inline cl/phase1/network/services/constants.go:32]: I am not sure but just curious, is this expiry time too short?
Does that mean we have only 24 secs to catch up block?

File: cl/phase1/forkchoice/fork_graph/fork_graph_disk.go [hunk 4]
```diff
@@ -361,31 +355,36 @@ func (f *forkGraphDisk) GetState(blockRoot libcommon.Hash, alwaysCopy bool) (*st
...
 		if !isSegmentPresent {
 			// check if it is in the header
 			bHeader, ok := f.GetHeader(currentIteratorRoot)
 			if ok && bHeader.Slot%dumpSlotFrequency == 0 {
-				break
+				copyReferencedState, err = f.readBeaconStateFromDisk(currentIteratorRoot)
+				if err != nil {
+					log.Trace("Could not retrieve state: Missing header", "missing", currentIteratorRoot, "err", err)
+				}
+				continue
```
[the remark this answers is not in the document]
Comment (Giulio2002) [reply cl/phase1/forkchoice/fork_graph/fork_graph_disk.go:371]: yes

File: cl/phase1/forkchoice/fork_graph/fork_graph_disk.go [hunk 5]
```diff
@@ -361,31 +355,36 @@ func (f *forkGraphDisk) GetState(blockRoot libcommon.Hash, alwaysCopy bool) (*st
...
 			return nil, nil
 		}
 		if block.Block.Slot%dumpSlotFrequency == 0 {
-			break
+			copyReferencedState, err = f.readBeaconStateFromDisk(currentIteratorRoot)
+			if err != nil {
+				log.Trace("Could not retrieve state: Missing header", "missing", currentIteratorRoot, "err", err)
+			}
+			if copyReferencedState != nil {
+				continue
```
[the remark this answers is not in the document]
Comment (Giulio2002) [reply cl/phase1/forkchoice/fork_graph/fork_graph_disk.go]: it is the same thing, but it is more readable i guess

File: cl/sentinel/gossip.go (hunk 1 above)
Comment (Giulio2002) [reply cl/sentinel/gossip.go:497]: Let's leave this for another PR

File: cl/sentinel/gossip.go (hunk 2 above)
Comment (Giulio2002) [reply cl/sentinel/gossip.go:211]: It is used in tests

File: cl/phase1/network/services/constants.go (hunk 3 above)
Comment (Giulio2002) [reply cl/phase1/network/services/constants.go:32]: It means that if we get a block from gossip and we dont have its parent, we memorize for 24 seconds, which would be 2 slots approximatelly. 7 minutes lead to high memory use, especially for blobs

**Review by domiwei — APPROVED** [review]

**Pull request merged**